### ISSUE TITLE
Finagle like Services & Filters

### ISSUE THREAD

**Issue by jamesanto** [zio/zio#1214]
I really like the concept of [**Services & Filters** from finagle](https://twitter.github.io/finagle/guide/ServicesAndFilters.html). My intention is to use the concept outside of finagle to model generic problems, such as AOP(using Filters). If we have something like this, it would help ensure that everyone thinks in terms of Services & Filters, which IMO, would result in better extensibility of code, maintainability & uniformity(of course I could be wrong). I was thinking if we can model them using ZIO. I came up with the following type aliases, but I am not very good at type variants and it doesn't work.

```scala
type Service[+Err, -Req, +Rep] = FunctionIO[Err, Req, Rep]
type Filter[+SErr, +FErr, -ReqIn, +RepOut, -ReqOut, -RepIn] = FunctionIO[FErr, (ReqIn, FunctionIO[FErr, ReqOut, RepIn]), RepOut]
type SimpleFilter0[+SErr, +FErr, -Req, +Rep] = Filter[SErr, FErr, Req, Rep, Req, Rep]
type SimpleFilter[+Err, -Req, +Rep] = Filter[Err, Err, Req, Rep, Req, Rep]
```

I could do something like the above with generic FlatMap type class. You can find it here : https://gist.github.com/jamesanto/971704b57e9a595d87525f16a83009fd. Even with this, I am not sure I got the type variants correct.

I am not even sure if we need something like this when we model our application using ZIO. Can we? If not is there a better way? Could you point me in the right direction?

**Comment by jdegoes** [comment]
Great idea!

The complexity of Finagle's `Filter` definition comes from the fact that `Future` is not pure.

Because ZIO effects are pure values, `Filter` can be simplified to:

```scala
type Filter[InErr, OutErr, InReq, InRep, OutReq, OutRep] = 
  Service[InErr, InReq, InRep] => Service[OutErr, OutReq, OutRep]
```

with appropriate variance.

Finally, there is no need to use `FunctionIO` since ZIO can now handle the same functions:

```scala
type Service[+Err, -Req, +Rep] = ZIO[Req, Err, Rep]
```

There you go, requests and filters in ZIO! 😄

**Comment by jamesanto** [comment]
Thank you so much!

Version 0.0.1 : https://gist.github.com/jamesanto/58b44fd630b502e95c8cccaadc27854d

**Comment by jdegoes** [comment]
@jamesanto Nice work! 😄  The other thing I'd do is add variance annotations (`+` / `-`) to the type parameters in `Filter`, as it will make inference work more smoothly.

**Comment by jamesanto** [comment]
I don't know how I got here, but only the following combination works. I went through the internet to try and understand variances and most of the examples explain the basics, which I can follow, but I am not able to apply it for filters(which explains that I don't really understand it fully).

```scala
type Service[+Err, -Req, +Rep] = ZIO[Req, Err, Rep]
  val Service: ZIO.type = ZIO
type Filter[-InErr, +OutErr, +InReq, -InRep, -OutReq, +OutRep] =
    Service[InErr, InReq, InRep] => Service[OutErr, OutReq, OutRep]
type SimpleFilter0[-InErr, +OutErr, Req, Rep] = Filter[InErr, OutErr, Req, Rep, Req, Rep]
type SimpleFilter[Err, Req, Rep] = SimpleFilter0[Err, Err, Req, Rep]
```

Since the compiler doesn't complain, I assume this is right, but I don't know if I will face issues down the line regarding inferencing.

ZIO an amazing work. A big thanks to all the contributors. I have been following this project for a long time now. I assume people will start using it more and more.

The library itself is well defined, but it would be more easy for people to adopt it once we have more examples and a set of best practices. The ecosystem page is great. I know there are works in progress for more documentation. A page on a set of opinionated-best-practices(TM :-D) would be great.
